### DISCUSSION TITLE
Preprocessing script for IEMOCAP (to solve: ValueError: Error: Number of utterances is not 5531...)

### DISCUSSION THREAD

**Discussion by bagustris** [speechbrain/speechbrain#2175] [Q&A] [answered]
Hi there, 

I am trying to run IEMOCAP recipe. Since the original dataset contains more than 10000 utterance and only parts of this dataset is used in SpeechBrain IEMOCAP recipe, I hit the following error when running the training script.

```
gus@pc-omen:emotion_recognition$ python3 train_with_wav2vec2.py hparams/train.yaml --data_folder /data/IEMOCAP_full_release/
speechbrain.core - Beginning experiment!
speechbrain.core - Experiment folder: results/ECAPA-TDNN/1968
speechbrain.core - Exception:
Traceback (most recent call last):
  File "/home/gus/github/speechbrain/recipes/IEMOCAP/emotion_recognition/train_with_wav2vec2.py", line 252, in <module>
    sb.utils.distributed.run_on_main(
  File "/home/gus/miniconda3/envs/nkululeko/lib/python3.9/site-packages/speechbrain/utils/distributed.py", line 61, in run_on_main
    func(*args, **kwargs)
  File "/home/gus/github/speechbrain/recipes/IEMOCAP/emotion_recognition/iemocap_prepare.py", line 79, in prepare_data
    raise ValueError(
ValueError: Error: Number of utterances is not 5531, please check your IEMOCAP folder
```

Is there a pre-processing script available to make 5531 utterance from the original dataset?

**Comment by BenoitWang** [comment]
Hi @bagustris, the pre-processing of the dataset is already integrated in `iemocap_prepare.py`, where you can find a [function](https://github.com/speechbrain/speechbrain/blob/16ef03604b187ff1d926368963bfda09515a47f0/recipes/IEMOCAP/emotion_recognition/iemocap_prepare.py#L290) that loads only the utterances annotated as `neu/sad/hap/exc/ang`. Since we never saw this issue before,  I suggest you verify your downloaded dataset maybe there are some audios missing. Is it from the official website? Or maybe try another source?

**Reply by bagustris** [reply]
Thanks @BenoitWang, for the answer.

Yes, it is from the official website. In my other research, I also used that number of utterances, but without using speechbrain.

It is strange that I got 5815 utterances instead of 5531 from `iemocap_prepare`. 
I inspected the directory for duplication (using fdupes), and there is no duplication.
Also, the total number of wav files from all sentences/wav is correctly shown as 10039, as in the original dataset.

**Reply by BenoitWang** [reply]
Hi, I share with you my json file for train+valid+test, you can download it [here](https://drive.google.com/file/d/1YdyYshWbeA66vRjBfQCH_UJlK0o4Bzp6/view?usp=sharing). There are 5531 items, so if you find the ones that exist in yours but not mine, we'll know where the problem comes from.

**Reply by bagustris** [reply, marked as answer]
Thanks @BenoitWang 

But It is strange again!

I got the exact number of utterances of 5531 same as you (checked by concatenating json files of train, valid, and test)  and it runs without error if I commented the following line in `iemocap_prepare.py`.

https://github.com/speechbrain/speechbrain/blob/29eb7429046a44394e08e29b59545e6b6caa1a4a/recipes/IEMOCAP/emotion_recognition/iemocap_prepare.py#L78-L81

into 
```
 # if sum([len(value) for value in speaker_dict.values()]) != NUMBER_UTT:
    #     raise ValueError(f"Error: Number of utterances is not 5531 but {sum([len(value) for value in speaker_dict.values()])}, please check your IEMOCAP folder"
```

So, my current workaround is to comment on lines 78-81 above. Maybe I did editing in dataset files, but I can't remember it correctly.
